### PR TITLE
Interpolation for Function1D (float)

### PR DESCRIPTION
**Pull request by mkgessen**
# Interpolation for Function1D (float)

* This PR implements Cython classes for 1d interpolators. 
* Interpolators are used through interface class Interpolate1D (raysect/core/math/function/float/function1d/interpolate.pyx and interpolate.pxd)
* Interpolator1D class that allows usage of different interpolators/extrapolators
    * Interpolators: Linear, Cubic
    * Extrapolators: Nearest neighbour, Linear

* Unit tests can be found from: raysect/core/math/function/float/function1d/tests/test_interpolator_1d.py

### CODE REVIEW

File: raysect/core/math/function/float/function1d/interpolate.pyx [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+
+        if x.shape != f.shape:
+            raise ValueError(f'Shape mismatch between x array ({x.shape}) and f array ({f.shape}).')
+
+        # test monotonicity
+        if (np.diff(x) <= 0).any():
+            raise ValueError('The x array must be monotonically increasing.')
+
+        # create extrapolator per extrapolation_type argument
+        if extrapolator_idx == EXT_NONE:
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/interpolate.pyx:112]: Replace this with a dictionary lookup: { name: Class }

File: raysect/core/math/function/float/function1d/interpolate.pyx [hunk 2]
```diff
@@ -0,0 +1,538 @@
...
+        elif extrapolator_idx == EXT_NEAREST:
+            self._extrapolator = _Extrapolator1DNearest(x, f, extrapolation_range)
+        elif extrapolator_idx == EXT_LINEAR:
+            self._extrapolator = _Extrapolator1DLinear(x, f, extrapolation_range)
+        elif extrapolator_idx == EXT_QUADRATIC:
+            raise NotImplementedError("Quadratic extrapolator not implemented.")
+        else:
+            raise ValueError(f'Unsupported extrapolator type {extrapolation_type}.')
+
+        # create extrapolator per interpolation_type argument
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/interpolate.pyx:123]: As above, use a dictionary lookup.

File: raysect/core/math/function/float/function1d/interpolate.pyx [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+        Evaluates the interpolating function.
+
+        :param double x: the point for which an interpolated value is required
+        :return: the interpolated value at point x.
+        """
+        cdef int index = find_index(self._x, x)
+
+        if index == -1:
+            if x < self._x[0] - self._extrapolation_range:
+                raise ValueError(
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/interpolate.pyx:144]: Please write this on one line, it doesn't need to be split onto two. It would be perfectly legible on one.

File: raysect/core/math/function/float/function1d/interpolate.pyx [hunk 4]
```diff
@@ -0,0 +1,538 @@
...
+
+        cdef int interpolator_idx = _INTERPOLATION_TYPES.get(interpolation_type)
+
+        extrapolation_type = extrapolation_type.lower()
+        if extrapolation_type not in _EXTRAPOLATION_TYPES:
+            raise ValueError(f'Extrapolation type {extrapolation_type} not found.')
+
+        cdef int extrapolator_idx = _EXTRAPOLATION_TYPES.get(extrapolation_type)
+
+        self.x = np.array(x, dtype=np.float64)
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/interpolate.pyx:90]: Allowing direct access to the data  means the user can modify the array which breaks the interpolation cache. For simplicity we assume that once an interpolator is instanced, the data it contains is immutable. A read-only numpy array should be returned (see the numpy api).

File: raysect/core/math/function/float/function1d/interpolate.pyx [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+                    f'The specified value (x={x}) is outside of extrapolation range.')
+            return self._extrapolator.evaluate(x, index)
+        else:
+            return self._interpolator.evaluate(x, index)
+
+    @property
+    def domain(self):
+        """
+        Returns bounding box of the provided inputs.
+        Order: min(x), max(x), min(f), max(f)
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/interpolate.pyx:159]: The min/max f is not required (and is not even valid for cubic interpolation). This is intended to be the sampling domain, so just the function input variable ranges.

In higher dimensions you'll want to group these in tuples, otherwise this will become a mess for higher dimensions objects. e.g. for 3D; ((xmin, xmax), (ymin, ymax), (zmin, zmax)). But for 1D, just (xmin, xmax) will do.

File: raysect/core/math/function/float/function1d/interpolate.pyx [hunk 6]
```diff
@@ -0,0 +1,538 @@
...
+
+    When called, stores cubic polynomial coefficients from the value of the function at the neighboring spline points
+    and the gradient at the neighbouring spline points based on central difference gradients. The polynomial
+    coefficients and gradients are calculated between each spline knots normalised to between 0 and 1.
+
+    :param x: 1D memory view of the spline point x positions.
+    :param f: 1D memory view of the function value at spline point x positions.
+    """
+    def __init__(self, double[::1] x, double[::1] f):
+        self._x = x
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/interpolate.pyx:206]: Storing the memory views is common to all the sub-classes, move to the base class.

File: raysect/core/math/function/float/function1d/interpolate.pyx [hunk 7]
```diff
@@ -0,0 +1,538 @@
...
+        self._n = n
+        # Where 'a' has been calculated already the mask value = 1
+        self._mask_a = np.zeros((n - 1,), dtype=np.float64)
+        self._a = np.zeros((n - 1, 4), dtype=np.float64)
+        self._a_mv = self._a
+
+    @cython.initializedcheck(False)
+    @cython.boundscheck(False)
+    @cython.cdivision(True)
+    cdef double get_gradient(self, double[::1] x_spline, double[::1] y_spline, int index):
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/interpolate.pyx:220]: Internal method: so use _get_gradient, not get_gradient.

I'd prefer this is renamed to `_calc_gradient`. As it is not just reading existing data, it is performing a calculation.

File: raysect/core/math/function/float/function1d/interpolate.pyx [hunk 8]
```diff
@@ -0,0 +1,538 @@
...
+                dfdx = (y_spline[index + 1]-y_spline[index - 1])/x_eff
+            else:
+                raise ZeroDivisionError('Two adjacent spline points have the same x value!')
+        return dfdx
+
+    @cython.cdivision(True)
+    @cython.boundscheck(False)
+    @cython.initializedcheck(False)
+    cdef double evaluate(self, double px, int index) except? -1e999:
+        cdef int index_use = find_index(self._x, px)
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/interpolate.pyx:261]: The Interpolate1D evaluate method already obtained the relevant array index (passed in `index`), this does not need to be repeated (it is also really expensive!).

File: raysect/core/math/function/float/function1d/interpolate.pyx [hunk 9]
```diff
@@ -0,0 +1,538 @@
...
+        # rescale x between 0 and 1
+        cdef double x_scal
+        cdef double[2] f, dfdx
+
+        cdef double x_bound = (self._x[index_use + 1] - self._x[index_use])
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/interpolate.pyx:266]: redundant brackets.

File: raysect/core/math/function/float/function1d/interpolate.pyx [hunk 10]
```diff
@@ -0,0 +1,538 @@
...
+                raise ZeroDivisionError('Two adjacent spline points have the same x value!')
+        return dfdx
+
+    @cython.cdivision(True)
+    @cython.boundscheck(False)
+    @cython.initializedcheck(False)
+    cdef double evaluate(self, double px, int index) except? -1e999:
+        cdef int index_use = find_index(self._x, px)
+        # rescale x between 0 and 1
+        cdef double x_scal
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/interpolate.pyx:263]: cdefs should all be defined at the top of the function, only inline cdef definitions in small, compact functions. It makes the code harder to read.

File: raysect/core/math/function/float/function1d/interpolate.pyx [hunk 11]
```diff
@@ -0,0 +1,538 @@
...
+        return self.get_gradient(self._x, self._f,  index_use)
+
+    def _test_evaluate_directly(self, x):
+        cdef int index = find_index(self._x, x)
+
+        """ Expose cython function for testing. Input the spline points x, f"""
+        return self.evaluate(x, index)
+
+
+cdef class _Interpolator1DCubicConstrained(_Interpolator1DCubic):
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/interpolate.pyx:311]: Remove this code until it is complete (move to a side branch)..

File: raysect/core/math/function/float/function1d/interpolate.pyx [hunk 12]
```diff
@@ -0,0 +1,538 @@
...
+    :param double extrapolation_range: Range covered by the extrapolator. Padded symmetrically to both ends of the input.
+    """
+    typename = NotImplemented
+
+    def __init__(self, double[::1] x, double[::1] f, double extrapolation_range):
+        self._range = extrapolation_range
+        self._x = x
+        self._f = f
+
+    cdef double extrapolate(self, double px, int order, int index, double rx) except? -1e999:
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/interpolate.pyx:410]: Why is this here? Looks like redundant code from a previous iteration?

File: raysect/core/math/function/float/function1d/interpolate.pyx [hunk 13]
```diff
@@ -0,0 +1,538 @@
...
+    :param object x: 1D array-like object of real values.
+    :param object f: 1D array-like object of real values.
+    :param double extrapolation_range: Range covered by the extrapolator.
+    """
+
+    typename = 'nearest'
+
+    def __init__(self, double [::1] x, double[::1] f, double extrapolation_range):
+        super().__init__(x, f, extrapolation_range)
+        self._last_index = self._x.shape[0] -1
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/interpolate.pyx:440]: Missing space after minus.

File: raysect/core/math/function/float/function1d/interpolate.pyx [hunk 14]
```diff
@@ -0,0 +1,538 @@
...
+
+    cdef double evaluate(self, double px, int index) except? -1e999:
+        if px < self._x[0]:
+            return self._f[0]
+        elif px >= self._x[self._last_index]:
+            return self._f[self._last_index]
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/interpolate.pyx:446]: Would be good form to add an exception if neither cases fire to indicate the method was given bad data,

File: raysect/core/math/function/float/function1d/interpolate.pyx [hunk 15]
```diff
@@ -0,0 +1,538 @@
...
+    :param object x: 1D array-like object of real values.
+    :param object f: 1D array-like object of real values.
+    :param double extrapolation_range: Range covered by the extrapolator.
+    """
+
+    typename = 'linear'
+
+    def __init__(self, double [::1] x, double[::1] f, double extrapolation_range):
+        super().__init__(x, f, extrapolation_range)
+        self._last_index = self._x.shape[0] -1
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/interpolate.pyx:461]: This is present in all extrapolators, move it to the base class.

File: raysect/core/math/function/float/function1d/interpolate.pyx [hunk 16]
```diff
@@ -0,0 +1,538 @@
...
+
+        if x.shape[0] <= 1:
+            raise ValueError(f'x array {np.shape(x)} must contain at least 2 spline points to linearly extrapolate.')
+
+    cdef double evaluate(self, double px, int index) except? -1e999:
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/interpolate.pyx:466]: Method will fail transparently if given an invalid index.

Comment (CnlPepper) [reply raysect/core/math/function/float/function1d/interpolate.pyx:466]: use if-elif-else, where else captures a bad index.

File: raysect/core/math/function/float/function1d/interpolate.pyx [hunk 17]
```diff
@@ -0,0 +1,538 @@
...
+        # The index returned from find_index is -1 at the array start or the length of the array at the end of array
+        if index == -1:
+            index += 1
+        else:
+            index -= 1
+        # Use a linear interpolator function to extrapolate instead
+        return lerp(self._x[index], self._x[index + 1], self._f[index], self._f[index + 1], px)
+
+
+cdef class _Extrapolator1DQuadratic(_Extrapolator1D):
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/interpolate.pyx:476]: Remove any code that is unsupported/not intended for release yet.

File: raysect/core/math/function/float/function1d/tests/test_interpolator_1d.py [hunk 18]
```diff
@@ -0,0 +1,349 @@
...
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/tests/test_interpolator_1d.py:31]: I think it would be easiest if we have a chat about the structure of the tests. I think this probably needs to be slightly restructured. Let me have a think about it....

File: raysect/core/math/cython/interpolation/cubic.pyx [hunk 19]
```diff
@@ -33,19 +33,19 @@ cimport cython
...
-        np.array([0.0, 5.0]),
-        np.array([2.0, 4.0]),
-        np.array([[0.0, 0.0], [1.0, 1.0]]),
-        np.array([[0.0, 0.0], [-1.0, -1.0]]),
-        np.array([[0.0, 0.0], [0.0, 0.0]]),
-        np.array([[0.0, 0.0], [0.0, 0.0]]),
-        x,
-        y
-    )
+#
```
Comment (CnlPepper) [inline raysect/core/math/cython/interpolation/cubic.pyx:36]: This should be removed.

File: raysect/core/math/function/float/function1d/interpolate.pxd [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+from raysect.core.math.function.float.function1d.base cimport Function1D
+from numpy cimport ndarray
+
+cdef class Interpolate1D(Function1D):
+    cdef:
+        ndarray x, f
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/interpolate.pxd:37]: Don't expose these arrays like this, see comment later in review.

File: raysect/core/math/function/float/function1d/interpolate.pxd [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+    cdef double get_gradient(self, double[::1] x_spline, double[::1] y_spline, int index)
+
+
+cdef class _Extrapolator1D:
+    cdef:
+        double _range
+        double [::1] _x, _f
+        int _last_index
+
+    cdef double extrapolate(self, double px, int order, int index, double rx) except? -1e999
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/interpolate.pxd:75]: Redundant code.

**Review by CnlPepper — CHANGES_REQUESTED** [review]
I've given it an initial code review. Generally good. There are some style inconsistencies and redundant code that need to be cleaned up/addressed. We should have a chat and briefly go over the comments. Feel free to start on making minor edits if you want, no point holding up minor changes.

**Comment by CnlPepper** [comment]
FYI I have not yet run the tests. The review was simply a code inspection so far. I want to think about the testing structure a bit as I think some aspect can be simpler, some aspects more refined.

File: raysect/core/math/function/float/function1d/interpolate.pyx (hunk 1 above)
Comment (CnlPepper) [reply raysect/core/math/function/float/function1d/interpolate.pyx:112]: Might be cleaner to add the name of each interpolator/extrapolator as a class attribute and then simply look them up e.g.:

```
cdef class MyExtrap:
    cdef readonly str ID = 'myextrap'

id_to_extrapolator = {
    MyExtrap.ID: MyExtrap
    ....
}

# get id string of any sub-object:
print(f'The strong id of this internal object is: {my_extrap.ID}')
```

Comment (mkgessen) [reply raysect/core/math/function/float/function1d/interpolate.pyx:112]: Agreed. However, Cython doesn't seem to support static member variables. 
I wonder if Python types would be feasible solution?

```cython
cdef class _Extrapolator1D:
    ID = NotImplemented

cdef class MyExtrap(_Extrapolator1D):
    ID = 'myextrap'

id_to_extrapolator = {
    MyExtrap.ID: MyExtrap
}
```

Alternatives that came to my mind:

1. Static methods

```cython
cdef class _Extrapolator1D:
 
   @staticmethod
     cdef str id():
           raise NotImplementedError('This is a base class..')

cdef class MyExtrap(_Extrapolator1D):
 
   @staticmethod
    cdef str id():
           return 'myextrap'

id_to_extrapolator = {
    MyExtrap.id(): MyExtrap
}
```

2. @classmethod + @cython.binding(True) instead of static

**Comment by mkgessen** [comment]
I believe all the reviewed points have been addressed now. In addition, this PR now includes 1D samplers + a very small unit test for them.

File: raysect/core/math/function/float/function1d/interpolate.pyx (hunk 3 above)
Comment (CnlPepper) [reply raysect/core/math/function/float/function1d/interpolate.pyx:128]: non essential style comment

File: raysect/core/math/function/float/function1d/tests/test_interpolator.py [hunk 22]
```diff
@@ -0,0 +1,222 @@
...
+            self.precalc_extrapolation = np.copy(self.precalc_extrapolation_nearest)
+        elif extrapolator_type == 'none':
+            self.precalc_extrapolation = None
+        else:
+            raise ValueError(
+                f'Extrapolation type {extrapolator_type} not found or no test. options are {id_to_extrapolator.keys()}'
+            )
+
+    def test_linear_interpolation(self):
+        self.setup_linear('none', 0.0)
```
Comment (CnlPepper) [inline raysect/core/math/function/float/function1d/tests/test_interpolator.py:175]: does not appear to actually run a test!

**Review by CnlPepper — APPROVED** [review]
Will merge into interpolation branch, however merging into development will wait until test work is completed and 2d/3d interpolators are added.
